### PR TITLE
test(lib/data): `ReplacementsSchema`

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Add JsonSchema and ZodSchema for the `replacements.json` file.

## Context
It will help users detect errors when adding new replacements instantly + he updated file will get validated against the zod schema during CI, leading to quicker PR merges & less reviews.
Ref: https://github.com/renovatebot/renovate/issues/29942#issuecomment-2241499559

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by secustor — COMMENTED** [review]
Can we generate the schema automatically with [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema)?

File: tools/schemas/replacements-schema.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (HonkingGoose) [inline tools/schemas/replacements-schema.json:2]:
```suggestion
  "$schema": "https://json-schema.org/draft-04/schema#",
```

Does the host support HTTPS? If yes, we should default to using HTTPS.

File: tools/schemas/schema.ts [hunk 2]
```diff
@@ -11,4 +11,46 @@ const MonorepoSchema = z.object({
...
+    .refine(
+      (rules) =>
+        rules.some(
+          (rule) =>
+            rule.replacementName !== undefined ||
+            rule.replacementNameTemplate !== undefined,
+        ),
+      {
+        message:
+          'At least one package rule must have either replacementName or replacementNameTemplate',
```
Comment (HonkingGoose) [inline tools/schemas/schema.ts:38]:
```suggestion
          'At least one package rule must use either the replacementName config option, or the replacementNameTemplate config option',
```

How about something like this?

File: lib/config/presets/internal/replacements.ts [hunk 3]
```diff
@@ -3,8 +3,10 @@ import type { Preset } from '../types';
 import type { PresetTemplate, Replacement } from './auto-generate-replacements';
 import { addPresets } from './auto-generate-replacements';
 
+const { $schema, ...tempPresets } = replacementGroups;
+
 /* eslint sort-keys: ["error", "asc", {"caseSensitive": false, "natural": true}] */
-export const presets: Record<string, Preset> = replacementGroups;
+export const presets: Record<string, Preset> = tempPresets;
```
Comment (viceice) [inline lib/config/presets/internal/replacements.ts:9]:
```suggestion
export const presets: Record<string, Preset> = replacementGroups;
delete presets['$schema'];
```

what about this? so we don't need an extra variable.

Comment (RahulGautamSingh) [reply lib/config/presets/internal/replacements.ts:9]: TS won't like it .. had to do the rpevious shenanigans for TS only

**Comment by RahulGautamSingh** [comment]
> Can we generate the schema automatically with [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema)?

Will it be useful? ie. can we use thus generated schemas for better UX?

**Comment by secustor** [comment]
> > Can we generate the schema automatically with [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema)?
> 
> Will it be useful? ie. can we use thus generated schemas for better UX?

We would not to have to manage a JSON and zod schema at the same time.

**Comment by RahulGautamSingh** [comment]
> We would not to have to manage a JSON and zod schema at the same time.

Got that.

For this we will need to add one step in the build process to save the schemas created using `zod-to-json` in json files, which will be referenced from the json files (`replacements.json`).

I think most users won't run the build process when adding a new replacement since its not necessary. So, maybe we are better to have a json schema. Also, these schemas won't change as much I guess so not a worry about maintaining.

File: tools/schemas/schema.ts [hunk 4]
```diff
@@ -11,4 +11,46 @@ const MonorepoSchema = z.object({
...
+    .refine(
+      (rules) =>
+        rules.some(
+          (rule) =>
+            rule.replacementName !== undefined ||
+            rule.replacementNameTemplate !== undefined,
+        ),
+      {
+        message:
+          'At least one package rule must use either the replacementName config option, or the replacementNameTemplate config option',
```
Comment (secustor) [inline tools/schemas/schema.ts:38]: Shouldn't all rules containing a `replacement` name.

**Review by secustor — COMMENTED** [review]
I, have meant that the json schema file still gets commited. 

Though lets keep it like that. I guess there will be not many changes to the schema anyway.

File: tools/schemas/schema.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply tools/schemas/schema.ts:38]: [There is one which does not.](https://github.com/renovatebot/renovate/blob/1562b303449c68e6529ad66a6924db5bf8a36e2a/lib/data/replacements.json#L183-L191) .

I am not sure it is needed here. But, decided to accomodate for it in in the Schema.

**Review by secustor — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the text. I'll let the maintainers check the code and technical working.

**Pull request merged**